docs/design/trimmed-assemblies.md

```diff
@@ -0,0 +1,215 @@
+
+```xml
+<ItemGroup>
+  <TrimmableAssembly Include="MyAssembly" />
+</ItemGroup>
+```
+
+The above will opt `MyAssembly.dll` into trimming. Note that the ItemGroup should contain assembly names without an extension, similar to [`TrimmerRootAssembly`](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options#root-assemblies). Before .NET 6 this would have been done with a target:
+
+```xml
+<Target Name="ConfigureTrimming"
+        BeforeTargets="PrepareForILLink">
+  <ItemGroup>
+    <ManagedAssemblyToLink Condition="'%(Filename)' == 'MyAssembly'">
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+## Future evolution
+
+As the .NET ecosystem shifts to support trimming of more libraries, we will be able to rely more on the trim analysis warnings to provide correctness guarantees. Since these guarantees are the same regardless of the `TrimMode`, we expect SDKs to move to more aggressive trimming defaults.
+
+We expect that the .NET SDK will eventually set `TrimMode` to `link` instead of `copyused` as it does today. Long-term, we may even go as far as enabling trimming of all assemblies by default when using ILLink. Existing MSBuild `IsTrimmable` and `TrimMode` metadata will continue to make it possible for the developer to opt in or out of trimming. We could additionally provide further knobs to simplify controlling trimming behavior and opting out of such defaults.
+
+### `TrimAllAssemblies` global opt-in
+
+We could make it it easier to enable trimming for all assemblies with a simple boolean. This would be equivalent to setting `IsTrimmable` to `true` on every assembly that is input to the linker. For example:
+
+```xml
+<PropertyGroup>
+  <TrimAllAssemblies>true</TrimAllAssemblies>
+</PropertyGroup>
+```
+
+could be used instead of
+
+```xml
+<Target Name="ConfigureTrimming"
+        BeforeTargets="PrepareForILLink">
+  <ItemGroup>
+    <TrimmerRootAssembly Include="@(IntermediateAssembly)" />
+    <ManagedAssemblyToLink>
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+This could be set by default in future SDKs, or it could be set by the developer in SDKs where it is not the default. We prefer not to introduce such a property at the moment, because it makes it too easy to enable this more "dangerous" behavior. We may consider adding this in the future when more of the .NET ecosystem has been made compatible with trimming.
+
+### `AssemblyMetadata("IsTrimmable", "False")`
+
+With more aggressive defaults, it could make sense to support an attribute opt-out via `[assembly: AssemblyMetadata("IsTrimmable", "False")]`. This would provide a way for developers to indicate that their assemblies should not be trimmed.
```

Comment:
For Xamarin.iOS it's fairly common thing to make the linker as aggressive as possible (to get the app size down), and then figure out how to unbreak everything that broke - by disabling the trimming for specific assemblies (or even give the linker xml definitions of what to preserve).

Scenario:

Customer has a class library project, and multiple app head projects (say one for Xamarin.iOS and Xamarin.TVOS) that reference the class library project. The customer tries to trim all assemblies, but realizes that the linker removes too much from his class library project, and wants to tell the linker to not trim his class library project. From this proposal, it seems that the customer has to modify the project file for each app head project (and any future projects that may end up referencing his class library project) to add the IsTrimmable metadata. Is that correct?

Reply:
I think this scenario should be sufficiently covered with `TrimmableAssembly`, so the developer would end up with something like this in the csproj file (assuming none of the assemblies have IsTrimmable set).

```xml
<ItemGroup>
  <TrimmableAssembly Include="assembly1" />
  <TrimmableAssembly Include="assembly2" />
  <TrimmableAssembly Include="assembly4" />
<!--  <TrimmableAssembly Include="assembly5" /> cannot be linked -->
  <TrimmableAssembly Include="assembly6" />
</ItemGroup>
```

Reply:
@marek-safar afaict that won't work if the developer sets `<TrimAllAssemblies>true</TrimAllAssemblies>` first.

Reply:
TrimAllAssemblies is not planned for .NET6, it's under Possible future work section https://github.com/mono/linker/pull/1647/files#diff-9abbbcd880aff3bdefcbefe9d49f810ae6eb372bf40816da53f3e470a8c483bfR116

Reply:
In the future if we had TrimAllAssemblies, we could add metadata to the TrimmableAssembly item to set Trimmable=false which would opt that assembly out of trimming.

Reply:
The UX for this scenario is still not ideal. See https://github.com/dotnet/runtime/issues/45353 which just came in and is basically about a similar problem.

If all of these are NuGet packages it will be possible to inject custom properties/targets from the NuGet into the app's project to mark assemblies as non-trimmable. Ideally it should be possible for the author of the library which has trouble being trimmed to mark itself as non-trimmable. With the current design there are basically only two ways to do this:
* Using NuGet's MSBuild integration to set the IsTrimmable to false for the assembly (using custom targets)
* Using embedded descriptor to force root the entire assembly.

This proposal doesn't help with that scenario in any way really.

I honestly don't think it's in such a bad shape - the possible solution would be to enable `[AssemblyMetadataAttribute("IsTrimmable", "False")]` as described above. But there's basically no real difference between that and rooting the entire assembly via embedded descriptor. It's just not as convenient.

Reply:
@marek-safar there will be a way to enable linking for all assemblies (currently implemented with the `MtouchLink` MSBuild variable), so the point still stands: if linking is enabled for all assemblies, then your solution won't work.

Reply:
> For Xamarin.iOS it's fairly common thing to make the linker as aggressive as possible

Do we have actual telemetry for this is this more of a gut feel?

In .NET Native, we played with settings like this a lot and imagined how useful they are (offering various opt-in/opt-out granularities at assembly levels). We wrote blog posts and tools on how to tune the size.

Turned out we lived in a bubble where all of this was clear, easy to use, and useful, but the real world disagreed. Inspecting actual real world apps in the Windows Store we found out only around 3% of all apps in the Store ever touched any of the defaults.

Out of those 3%, pretty much everyone just asked to root more things because we were breaking their app. Only 0.03% of apps opted out of the defaults and asked for more aggressive trimming, even though our testing indicated 70% of apps could do that without seeing any trimming related bugs.

This was when Windows Phone was still a thing and size mattered more.

Reply:
> From this proposal, it seems that the customer has to modify the project file for each app head project (and any future projects that may end up referencing his class library project) to add the IsTrimmable metadata. Is that correct?

@rolfbjarne Just to confirm - that's correct. If aggressive linking is already common, I can see the usefulness of adding `AssemblyMetadataAttribute("IsTrimmable", "False")` in .NET 5, but I have some concerns:

In the scenario you outlined, is the desire to disable trimming only locally for the projects that reference it, or also for other developers? The attribute has the downside that it could be too permanent. If the attributed library ships as a nuget package, another developer who only uses a small portion of it would end up with a larger app. I added some notes along these lines to the doc.

> we could add metadata to the TrimmableAssembly item to set Trimmable=false

@eerhardt I added this suggestion as an alternative under `NonTrimmableAssembly` - but I don't think this would address @rolfbjarne's concern since it still needs to be set in each app's project file.

> Using embedded descriptor to force root the entire assembly.

@vitek-karas with lazy loading, this will only keep the assembly if some part of it is used in a way that's understood by the linker (`IsTrimmable=false` metadata/attribute would keep it regardless). This may or may not be the right behavior, depending on @rolfbjarne's scenario.

Reply:
@MichalStrehovsky 

> > For Xamarin.iOS it's fairly common thing to make the linker as aggressive as possible
> 
> Do we have actual telemetry for this is this more of a gut feel?

I'll try to get some telemetry for this.

@sbomer 

> In the scenario you outlined, is the desire to disable trimming only locally for the projects that reference it, or also for other developers?

Also for other developers.

> The attribute has the downside that it could be too permanent. If the attributed library ships as a nuget package, another developer who only uses a small portion of it would end up with a larger app.

From this proposal it sounded like it would be possible to override the attribute with an MSBuild property, if the consumer of the library wanted to do so.

> > Using embedded descriptor to force root the entire assembly.
> 
> @vitek-karas with lazy loading, this will only keep the assembly if some part of it is used in a way that's understood by the linker (`IsTrimmable=false` metadata/attribute would keep it regardless). This may or may not be the right behavior, depending on @rolfbjarne's scenario.

Yes, this sounds like something that could work from a technical standpoint.

But from a developer perspective it just looks unpolished if I try to use `AssemblyMetadataAttribute("IsTrimmable", "False")` and then get a warning that either tells me to read some documentation that tells me to add a resource file with a specific name and content to accomplish more or less the same thing as what `AssemblyMetadataAttribute("IsTrimmable", "False")` would do.

Reply:
>> The attribute has the downside that it could be too permanent.[...]
>
> From this proposal it sounded like it would be possible to override the attribute with an MSBuild property

True, the MSBuild overrides make any combination possible - I'm just worried about the defaults you get from the attributes, especially as more people start to use aggressive trimming.

I added a description of your scenario to the doc under the future section @rolfbjarne. I'll go ahead and merge this since I think there's agreement about the opt-in, and if there's more data about the opt-out we can revisit.